scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -201,108 +247,221 @@ def _highly_variable_genes_single_batch(
+    df.index.name = "gene"
+    return df
+
+
+def _get_mean_bins(
+    means: pd.Series, flavor: Literal["seurat", "cell_ranger"], n_bins: int
+) -> pd.Series:
+    if flavor == "seurat":
+        bins = n_bins
+    elif flavor == "cell_ranger":
+        bins = np.r_[-np.inf, np.percentile(means, np.arange(10, 105, 5)), np.inf]
     else:
         raise ValueError('`flavor` needs to be "seurat" or "cell_ranger"')
-    dispersion_norm = df["dispersions_norm"].to_numpy()
-    if n_top_genes is not None:
-        dispersion_norm = dispersion_norm[~np.isnan(dispersion_norm)]
-        dispersion_norm[
-            ::-1
-        ].sort()  # interestingly, np.argpartition is slightly slower
-        if n_top_genes > adata.n_vars:
-            logg.info("`n_top_genes` > `adata.n_var`, returning all genes.")
-            n_top_genes = adata.n_vars
-        if n_top_genes > dispersion_norm.size:
-            warnings.warn(
-                "`n_top_genes` > number of normalized dispersions, returning all genes with normalized dispersions.",
-                UserWarning,
-            )
-            n_top_genes = dispersion_norm.size
-        disp_cut_off = dispersion_norm[n_top_genes - 1]
-        gene_subset = np.nan_to_num(df["dispersions_norm"].to_numpy()) >= disp_cut_off
+
+    return pd.cut(means, bins=bins)
+
+
+def _stats_seurat(mean_bins: pd.Series, disp_grouped: SeriesGroupBy) -> pd.DataFrame:
+    """Compute mean and std dev per bin."""
+    disp_bin_stats = disp_grouped.agg(avg="mean", dev=partial(np.std, ddof=1))
+    # retrieve those genes that have nan std, these are the ones where
+    # only a single gene fell in the bin and implicitly set them to have
+    # a normalized disperion of 1
+    one_gene_per_bin = disp_bin_stats["dev"].isnull()
+    gen_indices = np.flatnonzero(one_gene_per_bin.loc[mean_bins])
+    if len(gen_indices) > 0:
         logg.debug(
-            f"the {n_top_genes} top genes correspond to a "
-            f"normalized dispersion cutoff of {disp_cut_off}"
+            f"Gene indices {gen_indices} fell into a single bin: their "
+            "normalized dispersion was set to 1.\n    "
+            "Decreasing `n_bins` will likely avoid this effect."
         )
-    else:
+        disp_bin_stats["dev"].loc[one_gene_per_bin] = disp_bin_stats["avg"].loc[
+            one_gene_per_bin
+        ]
+        disp_bin_stats["avg"].loc[one_gene_per_bin] = 0
+    return _unbin(disp_bin_stats, mean_bins)
+
+
+def _stats_cell_ranger(
+    mean_bins: pd.Series, disp_grouped: SeriesGroupBy
+) -> pd.DataFrame:
+    """Compute median and median absolute dev per bin."""
+    from statsmodels.robust import mad
+
+    with warnings.catch_warnings():
+        # MAD calculation raises the warning: "Mean of empty slice"
+        warnings.simplefilter("ignore", category=RuntimeWarning)
+        disp_bin_stats = disp_grouped.agg(avg="median", dev=mad)
+    return _unbin(disp_bin_stats, mean_bins)
```

Comment:
```suggestion
    return disp_bin_stats.loc[mean_bins].set_index(mean_bins.index)
```

This could be an easier to follow way to do this. It could also be easier to follow if argument naming more consistently followed the variable naming convention.

Reply:
done, in the aforementioned `_get_disp_stats` helper